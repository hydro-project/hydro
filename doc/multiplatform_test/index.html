<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="`multiplatform_test`"><title>multiplatform_test - Rust</title><script>if(window.location.protocol!=="file:")document.head.insertAdjacentHTML("beforeend","SourceSerif4-Regular-6b053e98.ttf.woff2,FiraSans-Italic-81dc35de.woff2,FiraSans-Regular-0fe48ade.woff2,FiraSans-MediumItalic-ccf7e434.woff2,FiraSans-Medium-e1aa3f0a.woff2,SourceCodePro-Regular-8badfe75.ttf.woff2,SourceCodePro-Semibold-aa29a496.ttf.woff2".split(",").map(f=>`<link rel="preload" as="font" type="font/woff2"href="../static.files/${f}">`).join(""))</script><link rel="stylesheet" href="../static.files/normalize-9960930a.css"><link rel="stylesheet" href="../static.files/rustdoc-e56847b5.css"><meta name="rustdoc-vars" data-root-path="../" data-static-root-path="../static.files/" data-current-crate="multiplatform_test" data-themes="" data-resource-suffix="" data-rustdoc-version="1.92.0-nightly (f6092f224 2025-09-22)" data-channel="nightly" data-search-js="search-e256b49e.js" data-stringdex-js="stringdex-061df703.js" data-settings-js="settings-c38705f0.js" ><script src="../static.files/storage-e2aeef58.js"></script><script defer src="../crates.js"></script><script defer src="../static.files/main-6dc2a7f3.js"></script><noscript><link rel="stylesheet" href="../static.files/noscript-263c88ec.css"></noscript><link rel="alternate icon" type="image/png" href="../static.files/favicon-32x32-eab170b8.png"><link rel="icon" type="image/svg+xml" href="../static.files/favicon-044be391.svg"></head><body class="rustdoc mod crate"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><rustdoc-topbar><h2><a href="#">Crate multiplatform_test</a></h2></rustdoc-topbar><nav class="sidebar"><div class="sidebar-crate"><h2><a href="../multiplatform_test/index.html">multiplatform_<wbr>test</a><span class="version">0.5.0</span></h2></div><div class="sidebar-elems"><ul class="block"><li><a id="all-types" href="all.html">All Items</a></li></ul><section id="rustdoc-toc"><h3><a href="#">Sections</a></h3><ul class="block top-toc"><li><a href="#multiplatform_test" title="`multiplatform_test`"><code>multiplatform_test</code></a><ul><li><a href="#installation" title="Installation">Installation</a></li><li><a href="#usage" title="Usage">Usage</a></li></ul></li></ul><h3><a href="#attributes">Crate Items</a></h3><ul class="block"><li><a href="#attributes" title="Attribute Macros">Attribute Macros</a></li></ul></section><div id="rustdoc-modnav"></div></div></nav><div class="sidebar-resizer" title="Drag to resize sidebar"></div><main><div class="width-limiter"><section id="main-content" class="content"><div class="main-heading"><h1>Crate <span>multiplatform_<wbr>test</span>&nbsp;<button id="copy-path" title="Copy item path to clipboard">Copy item path</button></h1><rustdoc-toolbar></rustdoc-toolbar><span class="sub-heading"><a class="src" href="../src/multiplatform_test/lib.rs.html#1-236">Source</a> </span></div><details class="toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><h2 id="multiplatform_test"><a class="doc-anchor" href="#multiplatform_test">§</a><code>multiplatform_test</code></h2>
<p>Provides a proc-macro that expands to testing on platforms relevant to
Hydroflow. By default, expands to testing on the host (using the normal
<code>#[test]</code> attribute) and wasm (using <code>#[wasm_bindgen_test]</code>).</p>
<p>For example, the test</p>

<div class="example-wrap"><pre class="rust rust-example-rendered"><code><span class="kw">use </span>multiplatform_test::multiplatform_test;

<span class="attr">#[multiplatform_test]
</span><span class="kw">fn </span>my_test() {
  <span class="comment">// ...
</span>}</code></pre></div>
<p>Expands to</p>

<div class="example-wrap"><pre class="rust rust-example-rendered"><code><span class="attr">#[test]
#[wasm_bindgen_test::wasm_bindgen_test]
</span><span class="kw">fn </span>my_test() {
  <span class="comment">// ...
</span>}</code></pre></div><h3 id="installation"><a class="doc-anchor" href="#installation">§</a>Installation</h3><div class="example-wrap"><pre class="language-toml"><code>[dependencies]
multiplatform_test = * # replace with version.</code></pre></div>
<p>If you’re using <code>wasm</code> naturally you will need to add the <a href="https://crates.io/crates/wasm-bindgen-test-macro/"><code>wasm_bindgen_test</code></a>
crate as a dependency.</p>
<h3 id="usage"><a class="doc-anchor" href="#usage">§</a>Usage</h3><h4 id="specifying-platforms"><a class="doc-anchor" href="#specifying-platforms">§</a>Specifying platforms</h4>
<p>There are many platforms which can be specified:</p>
<ul>
<li><code>test</code> - Adds a standard <a href="https://doc.rust-lang.org/reference/attributes/testing.html#the-test-attribute"><code>#[test]</code> attribute</a>.</li>
<li><code>tokio</code> - Adds a <a href="https://docs.rs/tokio/latest/tokio/attr.test.html"><code>#[tokio::test]</code> attribute</a>.</li>
<li><code>async_std</code> - Adds an <a href="https://docs.rs/async-std/latest/async_std/attr.test.html"><code>#[async_std::test]</code> attribute</a>.</li>
<li><code>hydroflow</code> - Adds a <a href="https://docs.rs/hydroflow/latest/hydroflow/attr.test.html"><code>#[dfir_rs::test]</code> attribute</a>.</li>
<li><code>wasm</code> - Adds a <a href="https://docs.rs/wasm-bindgen-test/0.3.36/wasm_bindgen_test/attr.wasm_bindgen_test.html"><code>#[wasm_bindgen_test::wasm_bindgen_test]</code> attribute</a>.</li>
<li><code>env_logging</code> - Registers <a href="https://docs.rs/env_logger/latest/env_logger/"><code>env_logger</code></a> for <a href="https://docs.rs/log/latest/log/"><code>log</code>ging</a>.</li>
<li><code>env_tracing</code> - Registers a <a href="https://docs.rs/tracing-subscriber/latest/tracing_subscriber/index.html#reexport.FmtSubscriber"><code>FmtSubscriber</code></a> with an <a href="https://docs.rs/tracing-subscriber/latest/tracing_subscriber/filter/struct.EnvFilter.html"><code>EnvFilter</code></a> for <a href="https://docs.rs/tracing/latest/tracing/"><code>tracing</code></a>.</li>
</ul>
<p>You can test on a subset of platforms by passing in the platforms in parens:</p>

<div class="example-wrap"><pre class="rust rust-example-rendered"><code><span class="kw">use </span>multiplatform_test::multiplatform_test;

<span class="attr">#[multiplatform_test(test, env_logging)]  </span><span class="comment">// Only test on the standard `#[test]` platform, but enables logging
</span><span class="kw">fn </span>my_test() {
  <span class="comment">// ...
</span>}</code></pre></div>
<p>expands to</p>

<div class="example-wrap"><pre class="rust rust-example-rendered"><code><span class="kw">use </span>multiplatform_test::multiplatform_test;

<span class="attr">#[test]
</span><span class="kw">fn </span>my_test() {
  <span class="kw">let _ </span>= env_logger::builder().is_test(<span class="bool-val">true</span>).try_init();
  <span class="comment">// ...
</span>}</code></pre></div></div></details><h2 id="attributes" class="section-header">Attribute Macros<a href="#attributes" class="anchor">§</a></h2><dl class="item-table"><dt><a class="attr" href="attr.multiplatform_test.html" title="attr multiplatform_test::multiplatform_test">multiplatform_<wbr>test</a></dt><dd>See the <a href="index.html" title="mod multiplatform_test">crate</a> docs for usage information.</dd></dl></section></div></main></body></html>