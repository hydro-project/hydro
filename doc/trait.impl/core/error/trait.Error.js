(function() {
    var implementors = Object.fromEntries([["hydro_cli",[["impl <a class=\"trait\" href=\"https://doc.rust-lang.org/nightly/core/error/trait.Error.html\" title=\"trait core::error::Error\">Error</a> for <a class=\"struct\" href=\"hydro_cli/struct.AnyhowError.html\" title=\"struct hydro_cli::AnyhowError\">AnyhowError</a>"]]]]);
    if (window.register_implementors) {
        window.register_implementors(implementors);
    } else {
        window.pending_implementors = implementors;
    }
})()
//{"start":57,"fragment_lengths":[282]}