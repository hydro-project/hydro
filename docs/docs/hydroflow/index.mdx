import {Redirect} from '@docusaurus/router';

<Redirect to="/docs/dfir" />
