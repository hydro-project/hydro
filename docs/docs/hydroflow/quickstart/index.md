# Quickstart
This section will get you up and running with Rust and Hydroflow, and work you through a set of concrete examples illustrating much of what you need to know to get productive with Hydroflow.
