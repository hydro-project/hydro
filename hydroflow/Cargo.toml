[package]
name = "hydroflow"
version = "0.1.0"
edition = "2021"

[features]
default = [ "async", "macros" ]
async = [ "futures" ]
macros = [ "hydroflow_macro", "hydroflow_datalog" ]

[[example]]
name = "echoserver"

[[example]]
name = "echo_serde_json"

[[example]]
name = "chat"

[[example]]
name = "graph_reachability"

[[example]]
name = "graph_unreachability"

[[example]]
name = "three_clique"

[[example]]
name = "two_pc"

[[example]]
name = "deadlock_detector"

[[example]]
name = "kvs"

[dependencies]
bincode = "1.3"
byteorder = "1.4.3"
bytes = "1.1.0"
futures = { version = "0.3", optional = true }
hydroflow_datalog = { optional = true, path = "../hydroflow_datalog" }
hydroflow_lang = { path = "../hydroflow_lang" }
hydroflow_macro = { optional = true, path = "../hydroflow_macro" }
pin-project-lite = "0.2.9"
pusherator = { path = "../pusherator" }
ref-cast = "1.0"
regex = "1"
sealed = "0.4"
serde = { version = "1", features = [ "derive" ] }
serde_json = "1"
slotmap = "1.0"
static_assertions = "1.1.0"
tokio = { version = "1.16", features = [ "full" ] }
tokio-stream = { version = "0.1.10", features = [ "io-util" ] }
tokio-util = { version = "0.7.4", features = [ "net", "codec" ] }
variadics = { path = "../variadics" }

[dev-dependencies]
chrono = { version = "0.4.20", features = [ "serde" ], default-features = false }
clap = { version = "3.0.14", features = [ "derive" ] }
colored = "2.0"
criterion = { version = "0.3", features = [ "async_tokio" ] }
futures = { version = "0.3" }
itertools = "0.10.3"
rand = "0.8.4"
textnonce = "1.0.0"
time = "0.3"
trybuild = "1.0.71"
hdrhistogram = "7"
zipf = "7.0.0"
core_affinity = "0.5.10"
