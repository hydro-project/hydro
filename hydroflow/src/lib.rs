#![feature(never_type)]
#![feature(type_alias_impl_trait)]
#![allow(type_alias_bounds)]
#![allow(clippy::let_and_return)]
#![allow(clippy::iter_with_drain)]
#![allow(clippy::explicit_auto_deref)]

pub mod builder;
pub mod compiled;
pub mod lang;
pub mod props;
pub mod scheduled;

pub use hydroflow_macro::*;

pub use pusherator;
pub use tokio;
pub use tuple_list::tuple_list as tl;
pub use tuple_list::tuple_list_type as tt;

#[cfg(doctest)]
mod booktest {
    macro_rules! booktest {
        ($i:ident) => {
            #[doc = include_str!(concat!("../../book/", stringify!($i), ".md"))]
            mod $i {}
        };
    }
    booktest!(example_1);
    booktest!(example_2);
    booktest!(example_3);
    booktest!(example_4);
    booktest!(example_5);

    booktest!(example_1_surface);
    booktest!(example_2_surface);
    booktest!(example_3_surface);
    booktest!(example_4_surface);
    booktest!(example_5_surface);

    booktest!(surface_syntax);
    booktest!(surface_embedding);
    booktest!(surface_flows);
    booktest!(surface_data);
    booktest!(surface_ops);
}
