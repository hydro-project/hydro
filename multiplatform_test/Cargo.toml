[package]
name = "multiplatform_test"
publish = true
version = "0.0.0"
edition = "2021"
license = "Apache-2.0"
documentation = "https://docs.rs/multiplatform_test/"
description = "A simple attribute macro to combine `#[test]` and `#[wasm_bindgen_test]`"

[lib]
proc-macro = true

[dependencies]
proc-macro2 = "1.0"
quote = "1.0"
