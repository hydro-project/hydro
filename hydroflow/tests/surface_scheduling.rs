use std::error::Error;

use hydroflow::{assert_graphvis_snapshots, hydroflow_syntax, rassert_eq};
use multiplatform_test::multiplatform_test;

#[multiplatform_test(test, wasm, env_tracing)]
pub fn test_stratum_loop() {
    let (out_send, mut out_recv) = hydroflow::util::unbounded_channel::<usize>();

    let mut df = hydroflow_syntax! {
        source_iter([0]) -> union_tee;
        union_tee = union() -> tee();
        union_tee -> map(|n| n + 1) -> filter(|&n| n < 10) -> next_stratum() -> union_tee;
        union_tee -> for_each(|v| out_send.send(v).unwrap());
    };
    assert_graphvis_snapshots!(df);
    df.run_available();

    assert_eq!(
        &[0, 1, 2, 3, 4, 5, 6, 7, 8, 9],
        &*hydroflow::util::collect_ready::<Vec<_>, _>(&mut out_recv)
    );
    assert_eq!((11, 0), (df.current_tick(), df.current_stratum()));
}

#[multiplatform_test(test, wasm, env_tracing)]
pub fn test_tick_loop() {
    let (out_send, mut out_recv) = hydroflow::util::unbounded_channel::<usize>();

    let mut df = hydroflow_syntax! {
        source_iter([0]) -> union_tee;
        union_tee = union() -> tee();
        union_tee -> map(|n| n + 1) -> filter(|&n| n < 10) -> defer_tick() -> union_tee;
        union_tee -> for_each(|v| out_send.send(v).unwrap());
    };
    assert_graphvis_snapshots!(df);
    df.run_available();

    assert_eq!(
        &[0, 1, 2, 3, 4, 5, 6, 7, 8, 9],
        &*hydroflow::util::collect_ready::<Vec<_>, _>(&mut out_recv)
    );
    assert_eq!((10, 0), (df.current_tick(), df.current_stratum()));
}

#[multiplatform_test(hydroflow, env_tracing)]
async fn test_persist_stratum_run_available() -> Result<(), Box<dyn Error>> {
    let (out_send, out_recv) = hydroflow::util::unbounded_channel();

    let handle = tokio::task::spawn_blocking(|| {
        let mut df = hydroflow_syntax! {
            a = source_iter([0])
                -> persist()
                -> next_stratum()
                -> for_each(|x| out_send.send(x).unwrap());
        };
        assert_graphvis_snapshots!(df);
        df.run_available();
    });
    tokio::time::sleep(std::time::Duration::from_millis(200)).await;
    handle.abort();

    let seen: Vec<_> = hydroflow::util::collect_ready_async(out_recv).await;
    rassert_eq!(
        &[0],
        &*seen,
        "Only one tick should have run, actually ran {}",
        seen.len()
    )?;

    handle.await.unwrap();
    Ok(())
}

#[multiplatform_test(hydroflow, env_tracing)]
async fn test_persist_stratum_run_async() -> Result<(), Box<dyn Error>> {
    let (out_send, out_recv) = hydroflow::util::unbounded_channel();

    let mut df = hydroflow_syntax! {
        source_iter([0])
            -> persist()
            -> next_stratum()
            -> for_each(|x| out_send.send(x).unwrap());
    };
    assert_graphvis_snapshots!(df);

    tokio::time::timeout(std::time::Duration::from_millis(200), df.run_async())
        .await
        .expect_err("Expected time out");

    let seen: Vec<_> = hydroflow::util::collect_ready_async(out_recv).await;
    rassert_eq!(
        &[0],
        &*seen,
        "Only one tick should have run, actually ran {}",
        seen.len()
    )?;

    Ok(())
}
