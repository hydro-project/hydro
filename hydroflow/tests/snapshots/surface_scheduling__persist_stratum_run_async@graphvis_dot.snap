---
source: hydroflow/tests/surface_scheduling.rs
expression: df.meta_graph().unwrap().to_dot()
---
digraph {
    subgraph "cluster n1v1" {
        fillcolor="#dddddd"
        style=filled
        label = "sg_1v1\nstratum 0"
        n1v1 [label="(n1v1) source_iter([0])", fontname=Monaco, shape=invhouse, style = filled, color = "#0022ff", fontcolor = "#ffffff"]
        n2v1 [label="(n2v1) persist()", fontname=Monaco, shape=invhouse, style = filled, color = "#0022ff", fontcolor = "#ffffff"]
        n1v1 -> n2v1
    }
    subgraph "cluster n2v1" {
        fillcolor="#dddddd"
        style=filled
        label = "sg_2v1\nstratum 1"
        n3v1 [label="(n3v1) next_stratum()", fontname=Monaco, shape=invhouse, style = filled, color = "#0022ff", fontcolor = "#ffffff"]
        n4v1 [label="(n4v1) for_each(|x| out_send.send(x).unwrap())", fontname=Monaco, shape=house, style = filled, color = "#ffff00"]
        n3v1 -> n4v1
    }
    n2v1 -> n5v1
    n5v1 [label="(n5v1) handoff", fontname=Monaco, shape=parallelogram, style = filled, color = "#ddddff"]
    n5v1 -> n3v1 [arrowhead=box, color=red]
}

